CMakeLists.txt

```diff
@@ -1018,21 +1018,19 @@ configure_file(
 if(CONFIG_CPU_HAS_MPU AND CONFIG_USERSPACE)
 
     if(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT AND CONFIG_APP_SHARED_MEM)
-        set(GEN_APP_SMEM $ENV{ZEPHYR_BASE}/scripts/gen_app_smem.py)
+        set(GEN_APP_SMEM $ENV{ZEPHYR_BASE}/scripts/gen_app_partitions.py)
         set(APP_SMEM_LD "${PROJECT_BINARY_DIR}/include/generated/app_smem.ld")
-        set(OBJ_FILE_DIR "${PROJECT_BINARY_DIR}/../")
 
         add_custom_target(
           ${APP_SMEM_DEP} ALL
-          DEPENDS zephyr_prebuilt
          )
 
         add_custom_command(
           TARGET ${APP_SMEM_DEP}
           COMMAND ${PYTHON_EXECUTABLE} ${GEN_APP_SMEM}
-          -d ${OBJ_FILE_DIR}
+          -d ${APPLICATION_SOURCE_DIR}/src
```

Comment:
What if I put my source file in APPLICATION_SOURCE_DIR ?

Or I declared my partition in a header file that was included by an app source file?

I agree with @andyross [0], this approach does not work.

[0] https://github.com/zephyrproject-rtos/zephyr/pull/9288/files/0379cf9d6e50c83276361091cb67d6b40a948230#r208012638

Reply:
> What if I put my source file in APPLICATION_SOURCE_DIR 

Was an issue with the Cmakefile. The src was not needed. Fixed that.

> this approach does not work.

IMHO I don't agree with you.
I am modifying a merged code. This is how it was created. I am just fixing a issue inside it. 
I don't understand why this is a incorrect approach, when that was merged in the first place.
If you think that this is an incorrect approach then that PR should be reverted.

Reply:
What me and @andyross are objecting to is that we are now parsing C source code instead of object files.

This is a change brought by this PR.

I'm sure there is a bug in the original PR and it may be that reverting the original PR is the correct approach, I am not familiar with the details of the bug.

But the review feedback here is specifically that we are changing what the 'source' input to the generator is.

The problem with parsing C source code is that to do so reliably (e.g. users can write normal C code and expect it to build), you need a fully fledged preprocessor and/or compiler. This is not necessary however when one parses the binary output.

IMHO we have already brought up a sufficient amount of cases that demonstrate this:

Partitions created in header files outside of the 'app' are not picked up.

The macro does not behave like C any more, e.g. arguments must be literals, this is not required in 'normal' C, and therefore should not be required by the smem infrastructure.

Reply:
>What me and @andyross are objecting to is that we are now parsing C source code instead of object files.

If there was a way in which I could do it I would have done this by now. For this situation it is not possible at all. I created this patch as a last resort to salvage the issue.

>Partitions created in header files outside of the 'app' are not picked up.

This cant be done. This was not allowed in the previous implementation aswell. So as far as I can see this implementation is correct.

>The macro does not behave like C any more, e.g. arguments must be literals, this is not required in 'normal' C, and therefore should not be required by the smem infrastructure.

The current implementation takes the value provided in that line as a string and concatenates with other variables. This will be a string. I dont know why any user will need a MACRO to define a string which he will not use directly. 
And somethings we might need to enforce, like this will take only string as input. I don't see why we cant do this. To support an import feature we can make this rule.

Reply:
Are you able to prove that it can't be done?

Or at the very least, describe the problem in sufficient detail that it is clear to the reviewer that we have no other option than to parse C code.

Reply:
Please have a look at [0].

And i can prove that the current implementation doesnt work . The implementation you are suggesting.


0: https://github.com/zephyrproject-rtos/zephyr/pull/9288#issuecomment-411023245

Reply:
This is getting a bit out of hand.  Can you back up and explain exactly why the zephyr_prebuilt mechanism doesn't work for this application?  The whole point behind that extra step is so that we can perform tricks like this.  Why does segment alignment in the prebuilt binary (which gets thrown out) break things in the final link?

Assuming that's really the case and we have a chicken-and-egg bug in prelinkage: plausible ways to accomplish this without "grep for a macro":

+ Add a second prebuilt step in the linkage.  We used to do this for a brief moment in the pre-cmake world.   If you really have a piece of information that can only be derived from the linker **and** that is absolutely required to correctly link zephyr_prebuilt.elf, then this is really the only robust way to do it.

+ Scan the object files instead of the C code, though this probably requires a link step to get the sizes right.

+ Locate the files where that macro is expanded (may require some tricky magic) and run them through a second preprocessor-only step where the macro is expanded differently into easily detectable output.  We do something like this with xtensa to extract preprocessor symbols from their HAL layer.

But really, my intuition says that we should be able to make this work with one linker step.  Fundamentally all you're trying to do is get the size of a segment and then add some padding to align it.